File: packages/ts/src/json.ts
/** JSON serialization and file capture helpers. */
import { mkdir, writeFile, appendFile, access } from 'node:fs/promises';
import { constants } from 'node:fs';
import { dirname } from 'node:path';

export function dumpJson(value: unknown, opts: { ndjson?: boolean } = {}): string {
  if (opts.ndjson) {
    if (typeof value === 'string' || !value || typeof (value as Iterable<unknown>)[Symbol.iterator] !== 'function') {
      throw new TypeError('ndjson output requires an iterable of values');
    }
    return Array.from(value as Iterable<unknown>, (item) => JSON.stringify(item)).join('\n') + '\n';
  }
  return `${JSON.stringify(value, null, 2)}\n`;
}

export async function captureJson(
  value: unknown,
  path: string,
  opts: { ndjson?: boolean; overwrite?: boolean } = {},
): Promise<string> {
  if (!opts.overwrite) {
    try {
      await access(path, constants.F_OK);
      throw new Error(`Destination already exists: ${path}`);
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code !== 'ENOENT') throw err;
    }
  }
  await mkdir(dirname(path), { recursive: true });
  await writeFile(path, dumpJson(value, opts), 'utf8');
  return path;
}

/** Append one value as an NDJSON record to a capture created with captureJson. */
export async function appendJson(value: unknown, path: string): Promise<void> {
  await appendFile(path, dumpJson([value], { ndjson: true }), 'utf8');
}

File: packages/python/src/teleglance/json.py
"""JSON serialization and file capture helpers."""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from pathlib import Path
from typing import Any

from pydantic import BaseModel, TypeAdapter


def dump_json(value: Any, *, ndjson: bool = False) -> str:
    """Serialize models and JSON-compatible values for CLI or file output.

    ``ndjson`` expects an iterable and emits one compact JSON object per line.
    """
    if ndjson:
        if isinstance(value, (BaseModel, Mapping)):
            value = [value]
        elif isinstance(value, (str, bytes)) or not isinstance(value, Iterable):
            raise TypeError("ndjson output requires an iterable of values")
        return "".join(TypeAdapter(Any).dump_json(item).decode() + "\n" for item in value)
    return TypeAdapter(Any).dump_json(value, indent=2).decode() + "\n"


def capture_json(
    value: Any,
    path: str | Path,
    *,
    ndjson: bool = False,
    overwrite: bool = False,
) -> Path:
    """Serialize *value* to *path*, refusing to replace a file by default."""
    target = Path(path)
    if target.exists() and not overwrite:
        raise FileExistsError(f"Destination already exists: {target}")
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(dump_json(value, ndjson=ndjson), encoding="utf-8")
    return target
